### ISSUE TITLE
[java] New rule: NoEmptyUUID

### ISSUE THREAD

**Issue by Pankraz76** [pmd/pmd#6005]
**NoEmptyUUID:**

**Proposed Category:** One of [Error Prone]

**Description:**

UUID can never be empty

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

```java
UUID id = UUID.randomUUID();
boolean b = id.toString().isEmpty();
```

```java
UUID id = UUID.randomUUID();
boolean b = false;
```

**Possible Properties:**

*   Should this rule be customizable via properties?

**Comment by Pankraz76** [comment]
A clear rule to enforce better standards would help, so that everybody can relate.

**Comment by UncleOwen** [comment]
I don't think UUID.toString() ever returns an empty String. It's always a 36 character String that contains 4 dashes and 32 hex digits.

**Comment by Pankraz76** [comment]
yes exactly.


Imagine even for, so called, "senior architects", this is impossible to digest. So this rule need to tell them.

Ppl design whole systems but dont even understand the basics of some trivial random data type so called UUID but think, they have to overrule others, who know their 2 cents.

> It's always a 36 character String that contains 4 dashes and 32 hex digits.

Or `null`, as we live in silly stupid null-java-language.

**Comment by zbynek** [comment]
```
UUID id = UUID.randomUUID();
boolean b = id.toString().isEmpty();
```
is clearly wrong, but is it something that developers would actually write by mistake? The regression test found 0 occurences in all the checked projects: https://pull-requests.pmd-code.org/pr-6007/e81f3e99f95cb8ce1905fac6aeae4a488c4af41b/regression

**Comment by Pankraz76** [comment]
Not simple like that but had this condition in some if statement that never is positive, therefore this rule is in need.

**Comment by oowekyala** [comment]
This looks too specific to be a rule for general use. If we had a general dataflow analysis rule we could add this pattern to it, but it looks like a very unlikely mistake. Otherwise do we also need a rule for `anInteger.toString().isEmpty()`? And what about other toString that we know are never empty? Do we create one rule for each of them? Sounds like a slippery slope with limited usefulness.


> had this condition in some if statement that never is positive, therefore this rule is in need.

You could have found that out through coverage as well. Or you could write a rule for your specific project.

**Comment by Pankraz76** [comment]
both not gonna happen. Not going to enforce 100% coverage, not doing this. Yes its very silly and specific i see this. On the other hand its not doing harm right? If someone want to invest the runtime its up to them.

But yes i agree upon this is a very stupid mistake only seen once. As the team is now able to understand some rule for this would help. Assuming we are not alone.

**Comment by UncleOwen** [comment]
Yes, having silly rules causes harm. They cause maintenance overhead. They confuse users. They lower the reputation of the project. Can we just close this issue?
